src/serialization/serialization.rs

```diff
@@ -0,0 +1,412 @@
+extern crate byteorder;
+
+use super::{Counter, Histogram};
+use std::io::{self, Cursor, Read, Write};
+use std;
+use super::num::ToPrimitive;
+use self::byteorder::{BigEndian, ReadBytesExt, WriteBytesExt};
+
+#[path = "tests.rs"]
+#[cfg(test)]
+mod tests;
+
+#[path = "benchmarks.rs"]
+#[cfg(all(test, feature = "bench_private"))]
+mod benchmarks;
+
+const V2_COOKIE_BASE: u32 = 0x1c849303;
+
+const V2_COOKIE: u32 = V2_COOKIE_BASE | 0x10;
+
+const V2_HEADER_SIZE: usize = 40;
+
+/// Errors that occur during serialization.
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+pub enum SerializeError {
+    /// A count above i64::max_value() cannot be zig-zag encoded, and therefore cannot be
+    /// serialized.
+    CountNotSerializable,
+    /// Internal calculations cannot be represented in `usize`. Use smaller histograms or beefier
+    /// hardware.
+    UsizeTypeTooSmall,
+    /// An i/o operation failed.
+    IoError
+}
+
+impl std::convert::From<std::io::Error> for SerializeError {
+    fn from(_: std::io::Error) -> Self {
+        SerializeError::IoError
+    }
+}
+
+/// Serializer for the V2 binary format.
+/// Serializers are intended to be re-used for many histograms.
```

Comment:
Actually, repeated use is pretty much the default use-case I have: a bunch of histograms that all need to be periodically serialized and sent elsewhere for an analysis. I maintain https://bitbucket.org/marshallpierce/jersey2-metrics, for instance, which will keep a HdrHistogram for service time of each HTTP endpoint you have. That's the sort of thing I plan to do in Rust, as well. Consider https://github.com/Searchlight/khronus as well: a TSDB built on HdrHistogram. I think we should be viewing more problems through the lens of granular histograms that get aggregated or otherwise combined elsewhere for analysis.

Also, serialization is not a single concept. We already have four distinct version, of which one is implemented and the another will be soon. Would we have bring-your-own-vec versions of both, leading to 4 methods on `Histogram`? That feels sloppy to me. 

From a conceptual standpoint, I think serialization is a fundamentally distinct concern from recording. I think it is a mistake to put `save_to_db()` on your domain model objects/structs, and this feels like an analogous situation. I'd rather have people think about and decide what serialization they want than just see a `serialize()` and hope for the best, and I think an extra line is a reasonable syntactic burden.